keps/sig-node/4727-reasonable-image-gc/README.md

```diff
@@ -0,0 +1,439 @@
+to implement this enhancement.
+
+##### Prerequisite testing updates
+
+
+
+##### Unit tests
+
+
+- `pkg/kubelet/apis/config/validation`: `-:-:-` - `0%`
+
+##### Integration tests
+
+
+
+- <test>: <link to test coverage>
+
+##### e2e tests
+
+
+- <test>: <link to test coverage>
+
+### Graduation Criteria
+
+
+#### Alpha
+- Feature implemented behind feature gate.
+- Unit tests passed as designed in [TestPlan](#test-plan).
+
+#### Beta
+
+- Gather feedback from developers.
+
+#### GA
+- No negative feedback.
+- Update documents to reflect the changes.
+
+### Upgrade / Downgrade Strategy
+
+
+### Version Skew Strategy
+
+
+## Production Readiness Review Questionnaire
+
+
+### Feature Enablement and Rollback
+
+<!--
+This section must be completed when targeting alpha to a release.
+-->
+
+###### How can this feature be enabled / disabled in a live cluster?
+
+<!--
+Pick one of these and delete the rest.
+
+Documentation is available on [feature gate lifecycle] and expectations, as
+well as the [existing list] of feature gates.
+
+[feature gate lifecycle]: https://git.k8s.io/community/contributors/devel/sig-architecture/feature-gates.md
+[existing list]: https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/
+-->
+
+- [x] Feature gate (also fill in values in `kep.yaml`)
+  - Feature gate name: ImageGCHighThresholdAccurate
+  - Components depending on the feature gate: kubelet
+
+
+###### Does enabling the feature change any default behavior?
+
+Yes.
+
+###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?
+
+Yes, we can just disable the feature gate.
+
+###### What happens if we reenable the feature if it was previously rolled back?
+
+The constrains are respected again.
+
+###### Are there any tests for feature enablement/disablement?
+
+No.
+
+### Rollout, Upgrade and Rollback Planning
+
+
+
+###### How can a rollout or rollback fail? Can it impact already running workloads?
+
+It's an opt-in feature for end-users and will maintain current behaviors if not set, so
+it will not impact the running workloads.
+
+###### What specific metrics should inform a rollback?
+
+No.
+
+###### Were upgrade and rollback tested? Was the upgrade->downgrade->upgrade path tested?
+
+No.
+
+###### Is the rollout accompanied by any deprecations and/or removals of features, APIs, fields of API types, flags, etc.?
+
+No.
+
+### Monitoring Requirements
+
+No.
+
+###### How can an operator determine if the feature is in use by workloads?
+
+No.
+
+###### How can someone using this feature know that it is working for their instance?
+
+
+- [ ] Events
+  - Event Reason: 
+- [ ] API .status
+  - Condition name: 
+  - Other field: 
+- [ ] Other (treat as last resort)
+  - Details:
+
+###### What are the reasonable SLOs (Service Level Objectives) for the enhancement?
+
+
+
+###### What are the SLIs (Service Level Indicators) an operator can use to determine the health of the service?
+
+
+- [ ] Metrics
```

Comment:
should we introduce a new metric that signifies when the feature would be fatal? I think it'd be added regardless of the feature gate, and would emit when kubelet gc threshold is higher than eviction threshold

Reply:
basically: there should be a way for an admin to easily see when a set of nodes would fail upon setting this feature, to give time to prepare and update

Reply:
add `kubelet_image_gc_before_storage_eviction` metric denote whether `--image-gc-high-threshold` smaller than  `100 - imagefs.available`, which contains `image-gc-threshold` and `imagefs-available` labels

Reply:
what type is this metric?

Reply:
Gauge